### PR TITLE
[BIOMAGE-1036] - Run QC after GEM2S

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1036

#### Link to staging deployment URL 

in UI PR

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/295
API - https://github.com/biomage-ltd/api/pull/134

#### Anything else the reviewers should know about the changes here

- Moved saving processing settings from GEM2S into a pipeline hook
- Run QC automatically after GEM2S

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Review by kafkasl — CHANGES_REQUESTED** [review]
@aerlaut shouldn't there be tests also for the actual run method of the hooks instead of just the register?

**Review by kafkasl — APPROVED** [review]
[pebble] @aerlaut shouldn't there be tests also for the actual run method of the hooks instead of just the register?

File: src/utils/hookRunner.js [hunk 1]
```diff
@@ -15,7 +21,8 @@ class hookRunner {
       || this.hooks[taskName].length === 0
     ) { return null; }
 
-    // Manual looping is done to prevent passing function in hooks[taskName] into a callback
+    // Manual looping is done to prevent passing function in hooks[taskName] into a callback,
+    // which might cause scoping issues
     for (let idx = 0; idx < this.hooks[taskName].length; idx += 1) {
       if (this.hooks[taskName][idx].constructor.name === 'AsyncFunction') {
```
Comment (marcellp) [inline src/utils/hookRunner.js]: No need for this, you can `await` on a non-async function.

File: src/api/route-services/gem2s-response.js [hunk 2]
```diff
@@ -38,21 +42,14 @@ const gem2sResponse = async (io, message) => {
...
-  if (tableName.includes('experiments')) {
-    await experimentService.updateExperiment(experimentId, item);
-  } else if (tableName.includes('samples')) {
-    const { projectUuid } = item;
-    await samplesService.updateSamples(projectUuid, item);
-  }
+  pipelineHook.run(taskName, {
+    experimentId,
+    item,
+  });
```
Comment (marcellp) [inline src/api/route-services/gem2s-response.js]: this needs to be `await pipelineHook.run`

**Review by marcellp — CHANGES_REQUESTED** [review]

File: src/utils/hooks/clusteringWorkRequest.js [hunk 3]
```diff
@@ -2,23 +2,25 @@ const constants = require('../../api/general-services/pipeline-manage/constants'
 const workRequestBuilder = require('../workRequestBuilder');
```
Comment (StefanBabukov) [inline src/utils/hooks/clusteringWorkRequest.js:2]: There is already a folder 'customHooks' inside utils for this

Comment (aerlaut) [reply src/utils/hooks/clusteringWorkRequest.js:2]: Is that in the UI?

Comment (StefanBabukov) [inline src/utils/hooks/clusteringWorkRequest.js:2]: my bad yeah

File: src/api/route-services/gem2s-response.js (hunk 2 above)
Comment (aerlaut) [reply src/api/route-services/gem2s-response.js]: added

File: src/utils/hookRunner.js (hunk 1 above)
Comment (aerlaut) [reply src/utils/hookRunner.js]: removed

**Comment by aerlaut** [comment]
> [pebble] @aerlaut shouldn't there be tests also for the actual run method of the hooks instead of just the register?

It's there already, just cut off below the lines that are shown.

**Review by marcellp — APPROVED** [review]

**Pull request merged**